File: 자바스크립트/프린터.js
function solution(priorities, location) {
    const PRIOR = 0,
        INDEX = 1;
    const len = priorities.length;
    let maxNum = Math.max(...priorities)
    let que = priorities.map((prior, index) => [prior, index])

    while (que.length !== 0) {
        const item = que.shift()
        if (item[PRIOR] === maxNum) {
            if (item[INDEX] === location) return len - que.length;
            maxNum = Math.max(...(que.map((x) => x[0])))
        } else que.push(item);
    }

    return que;
}

function solution2(priorities, location) {
    let cnt = 0;
    let que = priorities.map((prior, index) => [index, prior])
    let priorNum = priorities.sort((a, b) => a - b)
    
    while (que.length > 0) {
        const [curLocation, curPrior] = que.shift();
        
        if (curPrior !== priorNum[priorNum.length-1]) {
            que.push([curLocation, curPrior])
            continue
        }
        
        if (curPrior === priorNum[priorNum.length-1]) {
            priorNum.pop()
            cnt++
        }
        
        if (curLocation === location) {
            return cnt
        }
    }

}

File: 파이썬/프린터.py
from collections import deque


def solution(priorities, location):
    N, PRIOR = len(priorities), 0
    que = deque(zip(priorities, range(N)))
    mx = max(que)[PRIOR]

    while que:
        if que[0][PRIOR] >= mx:
            prior, init_loc = que.popleft()
            if que:
                mx = max(que)[PRIOR]
            if init_loc == location:
                return N - len(que)
        else:
            que.append(que.popleft())
